File: iosApp/ShoppingListApp/GrpcAdapters.swift
import Foundation

/// gRPC adapter configuration for iOS.
/// Provides channel management and metadata injection.
struct GrpcConfig {
    // TODO: Configure from environment or Info.plist
    static var host: String = "localhost"
    static var port: Int = 50051
    static var useTLS: Bool = false
    static var defaultTimeoutSeconds: TimeInterval = 10
}

/// Metadata keys for gRPC headers.
enum GrpcMetadataKeys {
    static let authorization = "authorization"
    static let xRequestId = "x-request-id"
    static let idempotencyKey = "idempotency-key"
}

/// Protocol for providing authentication tokens.
protocol TokenProvider {
    func getAccessToken() async -> String?
}

/// Creates standard metadata headers for gRPC calls.
func createMetadata(
    accessToken: String? = nil,
    idempotencyKey: String? = nil
) -> [String: String] {
    var headers: [String: String] = [:]
    
    // Always add request ID for observability
    headers[GrpcMetadataKeys.xRequestId] = UUID().uuidString
    
    // Add authorization if available
    if let token = accessToken {
        headers[GrpcMetadataKeys.authorization] = "Bearer \(token)"
    }
    
    // Add idempotency key for mutations
    if let key = idempotencyKey {
        headers[GrpcMetadataKeys.idempotencyKey] = key
    }
    
    return headers
}

/// Result type for gRPC operations.
enum GrpcResult<T> {
    case success(T)
    case failure(GrpcError)
    
    var value: T? {
        if case .success(let value) = self {
            return value
        }
        return nil
    }
    
    var error: GrpcError? {
        if case .failure(let error) = self {
            return error
        }
        return nil
    }
}

/// Error types for gRPC operations.
enum GrpcError: Error {
    case unauthenticated(String)
    case permissionDenied(String)
    case notFound(String)
    case failedPrecondition(String)
    case unavailable(String)
    case deadlineExceeded(String)
    case unknown(String)
    
    var message: String {
        switch self {
        case .unauthenticated(let msg): return msg
        case .permissionDenied(let msg): return msg
        case .notFound(let msg): return msg
        case .failedPrecondition(let msg): return msg
        case .unavailable(let msg): return msg
        case .deadlineExceeded(let msg): return msg
        case .unknown(let msg): return msg
        }
    }
}

/// Base class for gRPC adapters with common functionality.
/// Subclasses implement specific service adapters.
class BaseGrpcAdapter {
    let tokenProvider: TokenProvider
    
    init(tokenProvider: TokenProvider) {
        self.tokenProvider = tokenProvider
    }
    
    /// Execute a gRPC call with standard headers and error handling.
    func executeCall<T>(
        idempotencyKey: String? = nil,
        block: @escaping ([String: String]) async throws -> T
    ) async -> GrpcResult<T> {
        do {
            let token = await tokenProvider.getAccessToken()
            let metadata = createMetadata(
                accessToken: token,
                idempotencyKey: idempotencyKey
            )
            let result = try await block(metadata)
            return .success(result)
        } catch {
            return .failure(mapError(error))
        }
    }
    
    /// Map errors to GrpcError type.
    private func mapError(_ error: Error) -> GrpcError {
        // TODO: Map specific gRPC status codes
        // For now, return unknown error
        return .unknown(error.localizedDescription)
    }
}

// MARK: - Placeholder Adapters

/// Placeholder for Auth gRPC adapter.
/// Will be implemented with actual gRPC client.
class AuthGrpcAdapter: BaseGrpcAdapter {
    // TODO: Implement AuthService RPC methods
}

/// Placeholder for Group gRPC adapter.
class GroupGrpcAdapter: BaseGrpcAdapter {
    // TODO: Implement GroupService RPC methods
}

/// Placeholder for List gRPC adapter.
class ListGrpcAdapter: BaseGrpcAdapter {
    // TODO: Implement ListService RPC methods
}

/// Placeholder for Sync gRPC adapter.
class SyncGrpcAdapter: BaseGrpcAdapter {
    // TODO: Implement SyncService RPC methods
}

File: androidApp/src/main/java/shopping/android/grpc/GrpcAdapters.kt
package shopping.android.grpc

import io.grpc.ManagedChannel
import io.grpc.ManagedChannelBuilder
import io.grpc.Metadata
import io.grpc.stub.MetadataUtils
import java.util.UUID
import java.util.concurrent.TimeUnit

/**
 * gRPC adapter configuration and channel management for Android.
 * Provides interceptors for authorization, request IDs, and idempotency keys.
 */
object GrpcConfig {
    // TODO: Configure from build config or environment
    var host: String = "10.0.2.2" // Android emulator localhost
    var port: Int = 50051
    var usePlaintext: Boolean = true
    var defaultTimeoutMs: Long = 10_000
}

/**
 * Creates and manages gRPC channels with proper interceptors.
 */
class GrpcChannelManager(
    private val tokenProvider: suspend () -> String?
) {
    private var channel: ManagedChannel? = null

    /**
     * Get or create the managed channel.
     */
    fun getChannel(): ManagedChannel {
        return channel ?: createChannel().also { channel = it }
    }

    /**
     * Create a new managed channel with interceptors.
     */
    private fun createChannel(): ManagedChannel {
        val builder = ManagedChannelBuilder
            .forAddress(GrpcConfig.host, GrpcConfig.port)

        if (GrpcConfig.usePlaintext) {
            builder.usePlaintext()
        }

        return builder.build()
    }

    /**
     * Shutdown the channel gracefully.
     */
    fun shutdown() {
        channel?.shutdown()?.awaitTermination(5, TimeUnit.SECONDS)
        channel = null
    }
}

/**
 * Metadata keys for gRPC headers.
 */
object GrpcMetadataKeys {
    val AUTHORIZATION: Metadata.Key<String> =
        Metadata.Key.of("authorization", Metadata.ASCII_STRING_MARSHALLER)

    val X_REQUEST_ID: Metadata.Key<String> =
        Metadata.Key.of("x-request-id", Metadata.ASCII_STRING_MARSHALLER)

    val IDEMPOTENCY_KEY: Metadata.Key<String> =
        Metadata.Key.of("idempotency-key", Metadata.ASCII_STRING_MARSHALLER)
}

/**
 * Creates metadata with standard headers.
 */
fun createMetadata(
    accessToken: String? = null,
    idempotencyKey: String? = null
): Metadata {
    val metadata = Metadata()

    // Always add request ID for observability
    metadata.put(GrpcMetadataKeys.X_REQUEST_ID, UUID.randomUUID().toString())

    // Add authorization if available
    accessToken?.let {
        metadata.put(GrpcMetadataKeys.AUTHORIZATION, "Bearer $it")
    }

    // Add idempotency key for mutations
    idempotencyKey?.let {
        metadata.put(GrpcMetadataKeys.IDEMPOTENCY_KEY, it)
    }

    return metadata
}

/**
 * Base class for gRPC adapters with common functionality.
 */
abstract class BaseGrpcAdapter(
    protected val channelManager: GrpcChannelManager,
    protected val tokenProvider: suspend () -> String?
) {
    protected val channel: ManagedChannel
        get() = channelManager.getChannel()

    /**
     * Execute a gRPC call with standard headers and error handling.
     */
    protected suspend inline fun <T> executeCall(
        idempotencyKey: String? = null,
        crossinline block: suspend (Metadata) -> T
    ): Result<T> {
        return try {
            val token = tokenProvider()
            val metadata = createMetadata(
                accessToken = token,
                idempotencyKey = idempotencyKey
            )
            Result.success(block(metadata))
        } catch (e: Exception) {
            Result.failure(mapGrpcException(e))
        }
    }

    /**
     * Map gRPC exceptions to domain errors.
     */
    protected fun mapGrpcException(e: Exception): Exception {
        // TODO: Map specific gRPC status codes to domain errors
        // io.grpc.StatusException or io.grpc.StatusRuntimeException
        return e
    }
}
